src/policy/ancestor_packages.h

```diff
@@ -0,0 +1,167 @@
+ * FilteredAncestorSet() can be used to get a transaction's "subpackage," i.e. ancestor set within the
+ * package. Also excludes non-PACKAGE ones.
+ * */
+class AncestorPackage
+{
+    /** Whether m_txns contains a connected package in which all transactions are ancestors of the
+     * last transaction. This object is not aware of chainstate. So if m_txns only includes a
+     * grandparent and not the "connecting" parent, this will (incorrectly) determine that the
+     * grandparent is not an ancestor.
+     * Set in the constructor and then does not change, even if transactions are skipped.
+     * */
+    bool m_ancestor_package_shaped{false};
+
+    struct PackageEntry {
+        enum class State : uint8_t {
+            /** Default value.
+             * If calling LinearizeWithFees(), this tx must have a fee and vsize. */
+            PACKAGE,
+            /** Excluded from FilteredAncestor{Set,FeeAndVsize}.
+             * Should not appear in any transaction's ancestor_subset.
+             * Excluded in LinearizeWithFees().
+             * Set by calling IsIsInMempool() on this tx. */
+            MEMPOOL,
+            /** This and all descendant's FilteredAncestor{Set,FeeAndVsize} will be std::nullopt.
+             * Should not appear in any transaction's ancestor_subset.
+             * Excluded in LinearizeWithFees().
+             * Set by calling IsDanglingWithDescendants() on this tx or its ancestor. */
+            DANGLING,
+        // If a parent is DANGLING, all descendants must also be DANGLING.
+        // If a child is MEMPOOL, all ancestors should also be MEMPOOL.
+        };
+
+        CTransactionRef tx;
+
+        State m_state{State::PACKAGE};
+
+        /** This value starts as std::nullopt when we don't have any fee information yet. It can be
+         * updated by calling LinearizeWithFees() if this entry isn't MEMPOOL or DANGLING. */
+        std::optional<uint32_t> mining_sequence;
+
+        /** Fees of this transaction. Starts as std::nullopt, can be updated using AddFeeAndVsize(). */
+        std::optional<CAmount> fee;
+
+        /** Virtual size of this transaction (depends on policy, calculated externally). Starts as
+         * std::nullopt. Can be updated using AddFeeAndVsize(). */
+        std::optional<int64_t> vsize;
+
+        /** Txids of all PACKAGE ancestors. Populated in AncestorPackage ctor. If an ancestor
+         * becomes MEMPOOL or DANGLING, it is erased from this set. Use FilteredAncestorSet to get
+         * filtered ancestor sets. */
+        std::set<Txid> ancestor_subset;
+
+        /** Txids of all in-package descendant. Populated in AncestorPackage ctor and does not
+         * change within AncestorPackage lifetime. */
+        std::set<Txid> descendant_subset;
+
+        explicit PackageEntry(CTransactionRef tx_in) : tx(tx_in) {}
+
+        // Used to sort the result of Txns, FilteredTxns, and FilteredAncestorSet. Always guarantees
```

Comment:
If my understanding of the guarantees of the `PackageEntry` comparison operator is correct, a breakage of the guarantee would be a yield `AncestorPackage` that produce a consensus-invalid block template, e.g a more incentive-compatible subset of descendants where the parents are sorted after. This comment to understand how test coverage should be built for this PR.

Reply:
If the sorting is broken, it means that `FilteredTxns` etc. might return transactions out of order, but this means validation will get a "missing inputs" error and nothing will be submitted. It can also mean that `FilteredTxns` doesn't give the more incentive compatible transactions first, which can affect whether we accept or reject these transactions.

This function being incorrect can't make us produce a consensus-invalid block template. Transactions still need to pass validation to be submitted, `CTxMemPool` sorts its entries, and `BlockAssembler` does its own sanity checks.